### PR TITLE
Add qmake support. Closes #12.

### LINKED ISSUE #12: qmake support
It'd be nice to have support for [qmake](http://doc.qt.io/qt-5/qmake-manual.html).

I gave it a quick try, but I got stuck trying to set a custom compiler for qmake... Apparently it'd involve writing a custom [mkspec](http://doc.qt.io/qt-5/qmake-environment-reference.html#qmakespec). Bleh.

### PR DESCRIPTION
**Pull request by The-Compiler**

### CODE REVIEW

File: config_gen.py [hunk 1]
```diff
@@ -162,6 +163,7 @@ def fake_build(project_dir, c_build_log_path, cxx_build_log_path, verbose, make_
     env_config = env.copy()
     env_config["YCM_CONFIG_GEN_CC_PASSTHROUGH"] = cc
     env_config["YCM_CONFIG_GEN_CXX_PASSTHROUGH"] = cxx
+    env_config["QMAKESPEC"] = "linux-clang"
```
Comment (rdnetto) [inline config_gen.py]: This should be performed within the if-case for qmake. (Similar to how the cmake case sets `cwd`.)

Also, I'm almost certain this is set to the wrong value, since the fake compiler uses `clang` and `clang++` instead.

File: config_gen.py [hunk 2]
```diff
@@ -216,6 +218,20 @@ def run(cmd, *args, **kwargs):
         else:
             run([make_cmd, "maintainer-clean"], env=env, **proc_opts)
 
+    elif(glob.glob(os.path.join(project_dir, "*.pro"))):
+        # qmake
```
Comment (rdnetto) [inline config_gen.py:226]: This should be capitalized, for consistency.

File: README.md [hunk 3]
```diff
@@ -21,6 +21,7 @@ YCM-Generator can also be used as a Vim plugin. Once installed with Vundle/NeoBu
 * Supported build systems:
     + Make
     + CMake
+    + qmake
```
Comment (rdnetto) [inline README.md:24]: This should also be capitalised

File: config_gen.py (hunk 2 above)
Comment (The-Compiler) [inline config_gen.py:226]: I thought I'd use the upstream capitalization, but now I saw Autotools is capitalized as well - will do.

**Comment by rdnetto** [comment]
Additional questions:
- doesn't qmake support out-of-tree builds? If so, it would be preferable to use that since it would mean we don't need to run `make clean`. (See how cmake is handled for an example)
- can you please suggest a project that can be used to test this? (I should probably start documenting these somewhere...)
- what happens if the user has multiple `*.pro` files in the directory?

File: config_gen.py (hunk 1 above)
Comment (The-Compiler) [inline config_gen.py]: Makes sense. I think the value is right however:

```
florian@ws042:/usr/lib/x86_64-linux-gnu/qt5/mkspecs$ ls -d *clang*
linux-clang  linux-clang-libc++  macx-clang  macx-clang-32  macx-ios-clang
```

Note qmake is for the Qt project, which supports C++ only. The `-libc++` one would use `libc++` instead of `libstdc++` which is (I think) not what we want.

File: README.md (hunk 3 above)
Comment (The-Compiler) [inline README.md:24]: Here `Make` and `CMake` follow the upstream capitalization. Shouldn't `qmake` too? It's not spelled `Qmake` or `QMake` upstream anywhere (and it definitely looks wrong to me).

Comment (rdnetto) [inline README.md:24]: I'd prefer consistency over accuracy here, since the inconsistency looks ugly. Realistically, having them all being lowercase or all being capitalized are equally good, so I don't really mind which of those two you use. (And all being lowercase is probably more grammatically correct anyway)

**Comment by The-Compiler** [comment]
- out-of-tree builds: I'm not sure. Qt Creator (the Qt IDE) does that (calling it shadow builds), but I didn't see how to handle this with `qmake` in the manpage. I'll do some research.
- test project: Any of the [Qt examples](http://doc.qt.io/qt-5/all-examples.html) should do, but I can also post a small hello world example later if you'd like me to.
- multiple files: qmake will show the usage and exit with exit status 1. This causes the build to fail, which currently is displayed as a "ERROR: No commands were logged to the build logs" message. I think if any of the commands fail, the script should make this more obvious - but I might open a second PR for that if I find the time, I don't think it's relevant here.

File: config_gen.py (hunk 1 above)
Comment (rdnetto) [inline config_gen.py]: Ok, looks like I misunderstand what it was for.

There are contexts in which one might want to use libc++ instead of libstdc++ on Linux (primarily relating to using Clang for C++11/C++14 support without having GCC 4.9 installed), but libstdc++ is the default and I'm fine with letting anyone who wants to vary from that figure it out themselves.

**Comment by rdnetto** [comment]
The problem with the Qt examples is I don't see any easy way to download all of the files in a single tarball/zipfile. The other problem is that they're very much toy examples - a real project with more complexity would be a better test case, because it's more likely to catch edge cases. e.g. I used the Linux kernel to test the Makefile logic, because Kbuild is easily the most complex set of Makefiles in existence (that I know of), and does things like call make recursively. (I'm aware this contradicts my earlier suggestion that it should ideally be small, but an accurate test is better than a fast one.)

The note about qmake failing if multiple files are detected should probably be included in a comment then. (Or even better, explicitly check for it and display an error message.)

I agree that the more information should probably be displayed for the error case - that could probably be done easily enough since everything uses `run()` already.

**Comment by The-Compiler** [comment]
The first thing coming to mind is [Otter Browser](https://github.com/OtterBrowser/otter-browser/) - they primarily use CMake but also provide qmake project files, so that could perhaps also be used to see if the output differs.

I'll let you know when I updated the PR, but I don't know if I'll get to it this week.

**Comment by rdnetto** [comment]
The problem we're going to run into is that most projects use a proper build system like make, etc. to invoke qmake. So for example, Otter uses cmake, and that will take precedence over the `.pro` file. I suspect the best solution is probably to just delete the cmake file for testing purposes - it's a hack, but it's probably the best option.

Feel free to take your with the PR - I just did a stable release a few days back, so I'm not planning on doing one any time soon (and even if I were, it's not like I have a formal schedule or anything...)

**Comment by The-Compiler** [comment]
> The problem we're going to run into is that most projects use a proper build system like make, etc. to invoke qmake.

That sounds backwards. qmake is generating a `Makefile` so `make` can run, similar to autotools/CMake.

Otter is just using CMake as an alternative to qmake.

**Comment by rdnetto** [comment]
> That sounds backwards. qmake is generating a Makefile so make can run, similar to autotools/CMake.

Sorry, I should have used `cmake` there instead of `make`. I _thought_ that cmake was just invoking qmake, but now that I look at it in detail it looks like `qmake` does the exact same thing as `cmake`, so just ignore what I said (except for the bit about deleting the cmake file).

**Comment by The-Compiler** [comment]
I've changed the capitalization for everything to lower-case and changed the build to be out-of-tree.
I also added a check so only one `.pro` file can exist.

Somehow things don't seem to work with OtterBrowser (not even with CMake), so I tried with [ExtPlane](https://github.com/vranki/ExtPlane) which also uses external libraries, and it seems to work fine.

Currently I duplicate the `glob.glob(os.path.join(project_dir, "*.pro"))` in the `elif` and in the body - if you prefer to have this only once (and above the whole big if-block), let me know.

Other than that, it looks ready for me - let me know if there's anything else!

**Comment by rdnetto** [comment]
> Currently I duplicate the `glob.glob(os.path.join(project_dir, "*.pro"))` in the elif and in the body - if you prefer to have this only once (and above the whole big if-block), let me know.

Yeah, I think that would be best. You're already assigning it to a variable, so might as well define it above with a comment to say it's used for qmake (similar and next to make_args).

I tried running YCM-Gen against ExtPlane, and it fails with:

```
$ qmake /tmp/mnt/ExtPlane/extplane-plugin.pro
Could not find mkspecs for your QMAKESPEC(linux-clang) after trying:
    /usr/share/qt4/mkspecs
Error processing project file: /tmp/mnt/ExtPlane/extplane-plugin.pro
```

It looks like the correct value/path for `QMAKESPEC` on my system is `unsupported/linux-clang`. (I'm running Sabayon, which is a Gentoo derivative.) My mkspec path is `/usr/share/qt4/mkspecs` - if it's the same for you then we can just put an if-statement in there to check which one exists. (Actually, you could probably do it more concisely with a functional approach using `filter`.)

**Comment by The-Compiler** [comment]
It seems your qmake is a Qt4 rather than a Qt5 one - since Qt4 support [ends in December](http://blog.qt.io/blog/2014/11/27/qt-4-8-x-support-to-be-extended-for-another-year/) and many projects (like ExtPlane as well) support Qt5, I don't think Qt4 should be supported.

I don't really know how you'd select what Qt to use on Gentoo/Sabayon. For Debian, there's the `QT_SELECT` environment variable. Though I guess it should be pointed out in the docs that it's Qt5 only.

**Comment by rdnetto** [comment]
`QT_SELECT` works in Sabayon too. We should probably add it to `env_config` to make it explicit.

Given that Qt4 is still supported, and there are plenty of projects which still target it (and will be even after it's no longer supported), I'd prefer to support it (especially since porting a project to Qt5 is non-trivial). Even if the `.pro` files are the same, changing the version of qmake used will alter the include paths in the resulting config file.

I think the best way of handling this is probably to add a flag specifying the version (defaulting to 5), and to set `QT_SELECT` to its value. That way Qt5 is used by default, but users which require Qt4 can still use it.

**Comment by The-Compiler** [comment]
I added a `--qt-version {4,5}` argument now. It works with both Qt4 and Qt5 for ExtPlane for me.

I also moved the `glob.glob` so it's only called once.

**Comment by rdnetto** [comment]
Great, thanks for the PR.
